### PR TITLE
Clone cards while cloning boards

### PR DESCRIPTION
**Pull request by luka-nextcloud**
Signed-off-by: Luka Trovic <luka@nextcloud.com>

* Resolves: #3679
* Target version: master 

### Summary
Duplicating a board copies its lists, but not the card.

### Checklist

- [ ] Code is properly formatted
- [ ] Sign-off message is added to all commits
- [ ] Tests (unit, integration, api and/or acceptance) are included
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: lib/Service/BoardService.php [hunk 1]
```diff
@@ -673,15 +675,39 @@ public function clone($id, $userId) {
...
+		foreach ($cards as $card) {
+			$newCard = new Card();
+			$newCard->setTitle($card->getTitle());
+			$newCard->setStackId($insertedStack->getId());
+			$newCard->setType($card->getType());
+			$newCard->setOrder($card->getOrder());
+			$newCard->setOwner($card->getOwner());
+			$newCard->setDescription($card->getDescription());
+			$newCard->setDuedate($card->getDuedate());
+			$newCard = $this->cardMapper->insert($newCard);
```
Comment (vinicius73) [inline lib/Service/BoardService.php:701]: We can isolate this process in a method inside `Card` entity.

```suggestion
			$newCard = new Card();
			$newCard->setFromCard($card);
			$newCard->setStackId($insertedStack->getId());
			$newCard = $this->cardMapper->insert($newCard);
```
or
```suggestion
			$newCard = Card::copyFrom($card);
			$newCard->setStackId($insertedStack->getId());
			$newCard = $this->cardMapper->insert($newCard);
```

Comment (luka-nextcloud) [reply lib/Service/BoardService.php:701]: Yes. But Card is a RelationalEntity. Only accessors and mutators should be defined in it I think. I wanted to add it in CardService but CardService is already using BoardService so there is a dependency issue. What I think is to add a new layer or put it into CardMapper (but mappers are meant to be the ORM layer I think).

Comment (vinicius73) [reply lib/Service/BoardService.php:701]: I see, you are right
Looks like we need to think more about it to create a wise solution.

Comment (luka-nextcloud) [reply lib/Service/BoardService.php:701]: I think we can put the logic in a function just in case we need to isolate later on in a class.

**Comment by juliushaertl** [comment]
As discussed in the call, lets:
- [ ] Add a parameter to make the cloning of cards option in the backend
- [ ] Trigger a dialog on the clone action in the frontend
- [ ] Let the user decide if cards should be cloned as well or not (to still cover the use case of using a board as template)
- [ ] While the clone request is running show a nice loading/progress indicator
- [ ] Once done, show a done message and option to close or go to the cloned board

**Pull request closed without merging**